dev/ToastNotifications/ToastNotificationUtility.cpp

```diff
@@ -0,0 +1,192 @@
+        &bufferLength);
+
+    if (status == ERROR_FILE_NOT_FOUND)
+    {
+        GUID guidReference;
+        THROW_IF_FAILED(CoCreateGuid(&guidReference));
+
+        wil::unique_cotaskmem_string guidStr;
+        THROW_IF_FAILED(StringFromCLSID(guidReference, &guidStr));
+
+        std::wstring guidWideStr{ guidStr.get() };
+        RegisterValue(hKey, L"ToastGUID", reinterpret_cast<const BYTE*>(guidWideStr.c_str()), REG_SZ, guidWideStr.size() * sizeof(wchar_t));
+        return guidWideStr;
+    }
+    else
+    {
+        THROW_HR_IF(status, FAILED_WIN32(status));
+    }
+    return registeredGuidBuffer;
+}
+
+std::wstring RetrieveToastGuid()
+{
+    if (AppModel::Identity::IsPackagedProcess())
+    {
+        wchar_t appUserModelId[APPLICATION_USER_MODEL_ID_MAX_LENGTH] = {};
+        UINT32 appUserModelIdSize{ ARRAYSIZE(appUserModelId) };
+
+        THROW_IF_FAILED(GetCurrentApplicationUserModelId(&appUserModelIdSize, appUserModelId));
+        return appUserModelId;
+    }
+    else
+    {
+        return RetrieveUnpackagedToastGuid();
+    }
+}
+
+void RegisterAssets(std::wstring const& appId, winrt::hstring const& displayName, winrt::Uri const& iconUri, wil::unique_cotaskmem_string const& clsid)
+{
+    wil::unique_hkey hKey;
+    std::wstring subKey{ appIdentifierPath + appId };
+
+    THROW_IF_FAILED(RegCreateKeyEx(
+        HKEY_CURRENT_USER,
+        subKey.c_str(),
+        0,
+        nullptr,
+        REG_OPTION_NON_VOLATILE,
+        KEY_ALL_ACCESS,
+        nullptr,
+        &hKey,
+        nullptr));
+
+    RegisterValue(hKey, L"DisplayName", reinterpret_cast<const BYTE*>(displayName.c_str()), REG_EXPAND_SZ, displayName.size() * sizeof(wchar_t));
+    RegisterValue(hKey, L"IconUri", reinterpret_cast<const BYTE*>(iconUri.AbsoluteUri().c_str()), REG_EXPAND_SZ, iconUri.AbsoluteUri().size() * sizeof(wchar_t));
+
+    std::wstring wideStringClsid{ clsid.get() };
+    RegisterValue(hKey, L"CustomActivator", reinterpret_cast<const BYTE*>(wideStringClsid.c_str()), REG_SZ, wideStringClsid.size() * sizeof(wchar_t));
+
+    return;
+}
+
+void RegisterComServer(wil::unique_cotaskmem_string const& processName, wil::unique_cotaskmem_string const& clsid)
+{
+    wil::unique_hkey hKey;
+    std::wstring wideStringClsid = { clsid.get() };
+    std::wstring subKey{ clsIdPath + wideStringClsid + L"\\LocalServer32" };
+    THROW_IF_FAILED(RegCreateKeyEx(
+        HKEY_CURRENT_USER,
+        subKey.c_str(),
+        0,
+        nullptr,
+        REG_OPTION_NON_VOLATILE,
+        KEY_ALL_ACCESS,
+        nullptr,
+        &hKey,
+        nullptr));
+
+    std::wstring comRegistrationExeString{ L"\"" };
+    comRegistrationExeString.append(processName.get());
+    comRegistrationExeString.append(L"\" ");
+    comRegistrationExeString.append(L"----ToastActivated:");
+
+    RegisterValue(hKey, nullptr, reinterpret_cast<const BYTE*>(comRegistrationExeString.c_str()), REG_SZ, (comRegistrationExeString.size() * sizeof(wchar_t)));
```

Comment:
Have we tested this using wide-string characters (not ASCII)?

Reply:
Yup. This works.